### PR TITLE
Make DataArrays ready for 0.5

### PR DESCRIPTION
**Pull request by andreasnoack**
All kind of stuff: new indexing, new iterations, no functors, new `getfield` syntax for operators, string changes and some other deprecations. Tests are passing now on master but the string story is not completely settled yet in `Compat.jl` so we'll need a new tag there before tests pass on 0.4. It might also be the right time to stop 0.3 support. I don't want to spend much time on making the tests pass there.

I should also mention that some of the tests are now slower than they are on 0.4. However, it's difficult to say if this is because of my changes here or because of the general slowdown on master. I profiled the `reducedim` tests and there is spent a lot of time in `inference` (even after the first run) and I can figure out why.

### CODE REVIEW

File: src/broadcast.jl [hunk 1]
```diff
@@ -208,8 +208,8 @@ for bsig in (DataArray, PooledDataArray), asig in ((@compat Union{Array,BitArray
             B
         end
         # ambiguity
-        Base.map!(f::Base.Callable, B::$bsig, r::Range) =
-            invoke(Base.map!, (Base.Callable, $bsig, $asig), f, B, r)
+        # Base.map!(f::Base.Callable, B::$bsig, r::Range) =
```
Comment (nalimilan) [inline src/broadcast.jl]: If this isn't needed, why not remove it completely?

File: src/DataArrays.jl [hunk 2]
```diff
@@ -2,10 +2,14 @@ VERSION >= v"0.4.0-dev+6641" && __precompile__()
 
 module DataArrays
     using Base.Cartesian, Compat, Reexport
+    import Compat.String
     @reexport using StatsBase
 
     const DEFAULT_POOLED_REF_TYPE = UInt32
 
+    import Base: ==, !=, >, <, >=, <=, +, -, *, !, &, |, $, ^, /,
```
Comment (nalimilan) [inline src/DataArrays.jl]: Why do you import these and not e.g. `length`?

File: src/broadcast.jl (hunk 1 above)
Comment (andreasnoack) [inline src/broadcast.jl]: I simply forgot that. I'll remote it.

File: src/DataArrays.jl (hunk 2 above)
Comment (andreasnoack) [inline src/DataArrays.jl]: I only import the operators. I actually like to use the qualified method definition instead of importing the names but I (think I) had to do it this way because of the changes to `getfield`.

File: src/extras.jl [hunk 3]
```diff
@@ -98,4 +98,4 @@ function rep(x::AbstractVector; times::Integer = 1, each::Integer = 1)
     rep(x, times, each)
 end
 
-rep(x::Any, times::Integer) = fill(x, times)
+rep(x::Number, times::Integer) = fill(x, times)
```
Comment (nalimilan) [inline src/extras.jl]: Why is this needed?

File: src/indexing.jl [hunk 4]
```diff
@@ -186,25 +186,31 @@ end
     end
 end
 
+if VERSION > v"0.5-"
+    Base.unsafe_getindex(x::Number, i::Int) = (@inbounds r = x[i]; r)
```
Comment (nalimilan) [inline src/indexing.jl]: This looks weird. Why is is necessary?

Comment (andreasnoack) [inline src/indexing.jl]: The method was removed from Base in one of the indexing overhauls but it is used by one of the indexing methods in this package. It can probably be removed when somebody has the time to update the indexing methods here systematically.

File: src/extras.jl (hunk 3 above)
Comment (andreasnoack) [inline src/extras.jl]: It would clash with the definition at line 82. For some reason, I don't think it is detected on 0.4.

**Comment by andreasnoack** [comment]
@nalimilan Thanks for the comments. I've made the necessary changes for the 0.4 tests to pass and I've removed that 0.3 support since 0.3 doesn't have `@generated`.

Anybody knows why AppVeyor testing is enabled but no `.appveyor.yml` and who's in charge of AppVeyor for JuliaStats?

File: src/DataArrays.jl (hunk 2 above)
Comment (nalimilan) [inline src/DataArrays.jl]: OK.

File: src/indexing.jl (hunk 4 above)
Comment (nalimilan) [inline src/indexing.jl]: Indeed, that's what I thought. Maybe a FIXME would be useful here. I'd rather work on the replacement for DataArrays than work on improving it, but at least that would make issues visible.

File: test/reducedim.jl [hunk 5]
```diff
@@ -56,7 +56,7 @@ function safe_mapslices{T}(f::Function, A::AbstractArray{T}, region, skipna)
         idx[d] = 1:size(A,d)
     end
 
-    r1 = f(reshape(A[idx...], Asliceshape); skipna=skipna)
+    r1 = f(copy(reshape(A[idx...], Asliceshape)); skipna=skipna)
```
Comment (nalimilan) [inline test/reducedim.jl]: Is a copy really necessary, since `r1` seems to be read-only?

**Comment by nalimilan** [comment]
Sorry, I was interrupted in the middle of the review. Thanks for doing this. I agree dropping 0.3 is OK at this point.

Regarding AppVeyor, looks like we stopped in the middle of enabling it. See https://github.com/JuliaStats/DataArrays.jl/pull/175.

File: test/reducedim.jl (hunk 5 above)
Comment (andreasnoack) [inline test/reducedim.jl]: The reductions don't work with `ReshapedArray`s and the copy makes it a `DataArray`. The main point here, though, is that this is just a test function to make sure that the "real" reduction functions work correctly with respect to `NA`s.

**Comment by andreasnoack** [comment]
The are still a lot of packages to go through before we are ready for 0.5 so I'll probably merge this soon and start to look into `DataFrames`.

**Comment by johnmyleswhite** [comment]
LGTM

**Comment by nalimilan** [comment]
OK to merge then?

**Comment by andreasnoack** [comment]
Not yet. There are some ambiguities now that I'll have to resolve. I'm looking into it .

File: src/indexing.jl (hunk 4 above)
Comment (tkelman) [inline src/indexing.jl]: can this be removed yet?

Comment (ararslan) [inline src/indexing.jl]: @tkelman It's not on master, so it has been removed at some point

Comment (tkelman) [inline src/indexing.jl]: it's still there, just not version conditional any more

Comment (tkelman) [inline src/indexing.jl]: https://github.com/JuliaStats/DataArrays.jl/blob/v0.4.0/src/indexing.jl#L128

Comment (ararslan) [inline src/indexing.jl]: Oh yeah, I see now. https://github.com/JuliaStats/DataArrays.jl/pull/240
